### PR TITLE
Add a new `ExtractVariables` function to `compose/template` package

### PR DESCRIPTION
**Pull request by vdemeester**
It allows to get easily all the variables defined in a
composefile (the `map[string]interface{}` representation that
`loader.ParseYAML` returns at least) and their default value too.

This commit also does some small function extract on substitution
funcs to reduce a tiny bit duplication.

Can be really useful for project that enhance the compose implementation like `docker/app`.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

### CODE REVIEW

File: cli/compose/template/template.go [hunk 1]
```diff
@@ -93,25 +93,91 @@ func Substitute(template string, mapping Mapping) (string, error) {
...
+		if escaped := groups["escaped"]; escaped != "" {
+			return extractedValue{}, false
+		}
+		val := groups["named"]
+		if val == "" {
+			val = groups["braced"]
+		}
+		name := val
+		var defaultValue string
+		if strings.Contains(val, ":?") {
```
Comment (silvin-lubecki) [inline cli/compose/template/template.go]: nit
```go
switch{
case string.Contains(val, ":?"):
    name, _ = partition(val, ":?")
case strings.Contains(val, "?"):
    name, _ = partition(val, "?")
...
}
return extractedValue{name: name, value: defaultValue}, true
}
```

File: cli/compose/template/template_test.go [hunk 2]
```diff
@@ -172,3 +172,91 @@ func TestSubstituteWithCustomFunc(t *testing.T) {
...
+			dict: map[string]interface{}{
+				"foo": "${bar?foo}",
+			},
+			expected: map[string]string{
+				"bar": "",
+			},
+		},
+		{
+			dict: map[string]interface{}{
+				"foo": "${bar:-foo}",
```
Comment (thaJeztah) [inline cli/compose/template/template_test.go]: Nested substitutions are not supported currently? (e.g. `${foo:-$bar}`)

File: cli/compose/template/template_test.go [hunk 3]
```diff
@@ -172,3 +172,91 @@ func TestSubstituteWithCustomFunc(t *testing.T) {
...
+				},
+			},
+			expected: map[string]string{
+				"bar":   "foo",
+				"fruit": "banana",
+				"toto":  "",
+			},
+		},
+	}
+	for _, tc := range testCases {
```
Comment (thaJeztah) [inline cli/compose/template/template_test.go]: Could change this to subtests as a follow up?

Comment (vdemeester) [reply cli/compose/template/template_test.go]: Yes sir ! :stuck_out_tongue_closed_eyes:

File: cli/compose/template/template_test.go (hunk 2 above)
Comment (vdemeester) [reply cli/compose/template/template_test.go]: I'm not actually sure abou that :stuck_out_tongue_closed_eyes:

**Pull request merged**